### DISCUSSION TITLE
Mutating Actions during extraction

### DISCUSSION THREAD

**Discussion by e-cal** [AI-Planning/macq#59] [General]
Should the `Action` objects in a `Model` be the same object references from the `TraceList`, or copies?

Is it problem that the `Action` objects in `TraceList` get mutated by model extraction?

**Comment by haz** [comment]
This is actually a subtle and tough question to answer. I'm leaning towards having them be entirely different classes. At the very least, they should be separate copies.

The difficulty resides in the fact that trace generation will need action objects, _sometimes_, depending on how the traces were produced. And the observation tokens may not contain any action info (even worse -- the data may have come from something without action info as well!). The model extraction, on the other hand, has a requirement for very rich action expressiveness (we may get to the point of needing distributions over possible action effects on a single action object). So it _feels_ like they deserve different treatment.

Can you remind me during our macq meeting this week to chat about this with the full group?

**Reply by e-cal** [reply]
Will do!
